### DISCUSSION TITLE
`register()`ing multiple different callbacks?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#42] [Q&A] [answered]
Can register multiple different callbacks?

**Comment by GregoryConrad** [comment]
Would you mind explaining exactly what you are trying to achieve?

If you mean back-to-back side effects like this:
```dart
final foo = use.register((api) {/* ... */});
final bar = use.register((api) {/* ... */});
final foobar = use.someOtherSideEffect();
```
Then yes, under the hood this is exactly what is happening with the builtin side effects. The builtin side effects all form a tree with the leaves as `use.register`s (which are executed one after the other).

**Reply by busslina** [reply]
I return from the capsule a Record containing several Object (catched with `use.value`). I pass them the `api` to let them register dispose callbacks. So could be several dispose callback on the same `api`.

Will it work?

**Reply by GregoryConrad** [reply, marked as answer]
> So could be several dispose callback on the same api.

`registerDispose` multiple times on an API, or even with the same callback instance multiple times, is not a problem. However:

> I pass them the api to let them register dispose callbacks.

This couples your object with the `SideEffectApi`. It'd be better to have the object just have a `dispose()` function, and then call that dispose yourself with `use.effect`. I'd recommend doing it like this:

```dart
final myObj = use.lazyValue(MyObj.new);
use.effect(() => myObj.dispose, [myObj]); // will automatically clean up myObj on disposal
```

**Reply by busslina** [reply]
> This couples your object with the SideEffectApi.

I created this class `adhoc` to the capsule, but yeah, I will think to use `use.effect` (I never used it yet).

(*) I switched from `Records` of callbacks/data to `Records` of classes to handle better the complexity. But they are classes with the only purpose to handle the capsule. But yeah, `use.effect` seems more elegant.

**Reply by busslina** [reply]
```dart
final myObj = use.lazyValue(MyObj.new);
use.effect(() => myObj.dispose, [myObj]); // will automatically clean up myObj on disposal
```

The thing is that the object is immutable (`use.value`) so I think I can skip the dependencies part

**Reply by GregoryConrad** [reply]
> The thing is that the object is immutable (use.value) so I think I can skip the dependencies part

_Don't skip the dependencies_. Even if it is immutable and will be until the end of time, it is a best practice to put everything the closure captures into the dependencies.

**Reply by busslina** [reply]
Yes Sir!
